core/src/main/scala/kafka/controller/ReplicaStateMachine.scala

```diff
@@ -122,178 +139,214 @@ class ReplicaStateMachine(controller: KafkaController, stateChangeLogger: StateC
-                    logStateChange()
-                    false
-                  case None =>
-                    true
-                }
-              case None =>
-                true
-            }
-          if (leaderAndIsrIsEmpty && !controller.topicDeletionManager.isPartitionToBeDeleted(topicAndPartition))
-            throw new StateChangeFailedException(
-              s"Failed to change state of replica $replicaId for partition $topicAndPartition since the leader " +
-                s"and isr path in zookeeper is empty")
-
+          replicaState.put(replica, OnlineReplica)
+        }
+      case OfflineReplica =>
+        validReplicas.foreach { replica =>
+          controllerBrokerRequestBatch.addStopReplicaRequestForBrokers(Seq(replicaId), replica.topic, replica.partition, deletePartition = false, null)
+        }
+        val replicasToRemoveFromIsr = validReplicas.filter(replica => controllerContext.partitionLeadershipInfo.contains(replica.topicAndPartition))
+        val updatedLeaderIsrAndControllerEpochs = removeReplicasFromIsr(replicaId, replicasToRemoveFromIsr.map(_.topicAndPartition))
+        updatedLeaderIsrAndControllerEpochs.foreach { case (partition, leaderIsrAndControllerEpoch) =>
+          if (!topicDeletionManager.isPartitionToBeDeleted(partition)) {
+            val recipients = controllerContext.partitionReplicaAssignment(partition).filterNot(_ == replicaId)
+            controllerBrokerRequestBatch.addLeaderAndIsrRequestForBrokers(recipients,
+              partition.topic,
+              partition.partition,
+              leaderIsrAndControllerEpoch,
+              controllerContext.partitionReplicaAssignment(partition), isNew = false)
+          }
+          replicaState.put(PartitionAndReplica(partition.topic, partition.partition, replicaId), OfflineReplica)
+        }
+      case ReplicaDeletionStarted =>
+        validReplicas.foreach { replica =>
+          replicaState.put(replica, ReplicaDeletionStarted)
+          controllerBrokerRequestBatch.addStopReplicaRequestForBrokers(Seq(replicaId),
+            replica.topic,
+            replica.partition,
+            deletePartition = true,
+            callbacks.stopReplicaResponseCallback)
+        }
+      case ReplicaDeletionIneligible =>
+        validReplicas.foreach(replica => replicaState.put(replica, ReplicaDeletionIneligible))
+      case ReplicaDeletionSuccessful =>
+        validReplicas.foreach(replica => replicaState.put(replica, ReplicaDeletionSuccessful))
+      case NonExistentReplica =>
+        validReplicas.foreach { replica =>
+          val currentAssignedReplicas = controllerContext.partitionReplicaAssignment(replica.topicAndPartition)
+          controllerContext.partitionReplicaAssignment.put(replica.topicAndPartition, currentAssignedReplicas.filterNot(_ == replica.replica))
+          replicaState.remove(replica)
+        }
+    }
+  }
+  private def removeReplicasFromIsr(replicaId: Int, partitions: Seq[TopicAndPartition]) = {
```

Comment:
Add a new line above.

Could we explicitly define the return type of this method?